File: Sources/Gantry/Spoolbase/PhysicalSpoolStore.swift
import Foundation

/// Shared Spoolbase stores so the inventory window, the dashboard AMS popover and (later) the
/// consumption tracker all read/write one source of truth.
@MainActor
enum SpoolbaseShared {
    static let filaments = FilamentStore()
    static let spools = PhysicalSpoolStore()
}

/// Persistent store for physical spools and their consumption history. Mirrors `FilamentStore`'s
/// JSON-in-Application-Support pattern so the two live side by side. All mutations save immediately;
/// nothing critical is kept only in memory (spec §24).
@MainActor
final class PhysicalSpoolStore {
    private(set) var spools: [PhysicalSpool]
    private(set) var usageEvents: [SpoolUsageEvent]
    private let spoolsURL: URL
    private let usageURL: URL
    var onChange: (() -> Void)?

    init(spoolsURL: URL? = nil, usageURL: URL? = nil) {
        self.spoolsURL = spoolsURL ?? Self.defaultSpoolsURL
        self.usageURL = usageURL ?? Self.defaultUsageURL
        spools = Self.load([PhysicalSpool].self, from: self.spoolsURL) ?? []
        usageEvents = Self.load([SpoolUsageEvent].self, from: self.usageURL) ?? []
    }

    // MARK: Lookup

    func spool(id: String) -> PhysicalSpool? { spools.first { $0.id == id } }

    /// The spool currently sitting in a given physical slot (or nil if that slot is empty).
    func spool(at location: SpoolLocation) -> PhysicalSpool? {
        guard !location.isStorage else { return nil }
        return spools.first { $0.location.sameSlot(as: location) }
    }

    func spools(forDefinition definitionID: UUID) -> [PhysicalSpool] {
        spools.filter { $0.filamentDefinitionID == definitionID }
    }

    /// Next free short id like "SP-00001" (one past the highest existing number).
    func nextSpoolID() -> String {
        let maxNumber = spools.compactMap { spool -> Int? in
            guard spool.id.hasPrefix("SP-") else { return nil }
            return Int(spool.id.dropFirst(3))
        }.max() ?? 0
        return String(format: "SP-%05d", maxNumber + 1)
    }

    // MARK: Mutations

    func add(_ spool: PhysicalSpool) {
        spools.append(spool)
        changed()
    }

    /// Creates `count` fresh rolls of one filament definition, each with its own id, dropped straight
    /// into storage (spec §1: adding a filament to Spoolbase creates one physical roll per spool). A
    /// full roll has remaining == nominal == `weight`; an opened roll passes a smaller `remaining`.
    @discardableResult
    func createRolls(definitionID: UUID, count: Int, weight: Double, remaining: Double? = nil) -> [PhysicalSpool] {
        var created: [PhysicalSpool] = []
        for _ in 0..<max(0, count) {
            let rest = remaining ?? weight
            let spool = PhysicalSpool(id: nextSpoolID(), filamentDefinitionID: definitionID,
                                      nominalWeightGrams: weight, remainingWeightGrams: rest,
                                      status: rest < weight ? .active : .new, location: .storage,
                                      openedAt: rest < weight ? .now : nil)
            spools.append(spool)
            created.append(spool)
        }
        if !created.isEmpty { changed() }
        return created
    }

    func update(_ spool: PhysicalSpool) {
        guard let index = spools.firstIndex(where: { $0.id == spool.id }) else { return }
        var updated = spool
        updated.updatedAt = .now
        spools[index] = updated
        changed()
    }

    func delete(id: String) {
        spools.removeAll { $0.id == id }
        changed()
    }

    /// Manual correction of the remaining amount (spec §12: "Ustaw pozostałą ilość").
    func setRemaining(id: String, grams: Double) {
        guard let index = spools.firstIndex(where: { $0.id == id }) else { return }
        spools[index].remainingWeightGrams = max(0, min(grams, spools[index].nominalWeightGrams))
        spools[index].status = spools[index].remainingWeightGrams <= 0 ? .empty : spools[index].status
        spools[index].updatedAt = .now
        changed()
    }

    /// Manual weighing (spec §5 "Skoryguj wagę"): store a freshly measured net amount and stamp the
    /// weighing date. `tare` (empty-spool weight) is remembered so a later gross reading can be
    /// converted to net automatically. `grams` here is already net (gross minus tare).
    func correctWeight(id: String, netGrams: Double, tare: Double? = nil) {
        guard let index = spools.firstIndex(where: { $0.id == id }) else { return }
        spools[index].remainingWeightGrams = max(0, min(netGrams, spools[index].nominalWeightGrams))
        if let tare { spools[index].tareGrams = tare }
        spools[index].weighedAt = .now
        if spools[index].remainingWeightGrams <= 0 {
            spools[index].status = .empty
            spools[index].emptiedAt = .now
        } else if spools[index].status == .empty {
            spools[index].status = spools[index].location.isStorage ? .stored : .active
        }
        spools[index].updatedAt = .now
        changed()
    }

    /// Reset a roll back to a full nominal amount (spec §6): used when a spent roll is swapped for a
    /// fresh one of the same product without minting a new id. Clears consumption/lifecycle so the
    /// history starts over for the new physical roll. Pass `nominal` to also change the roll size.
    func resetToFull(id: String, nominal: Double? = nil) {
        guard let index = spools.firstIndex(where: { $0.id == id }) else { return }
        let full = max(0, nominal ?? spools[index].nominalWeightGrams)
        spools[index].nominalWeightGrams = full
        spools[index].remainingWeightGrams = full
        spools[index].totalConsumedGrams = 0
        spools[index].openedAt = nil
        spools[index].emptiedAt = nil
        spools[index].weighedAt = nil
        spools[index].status = spools[index].location.isStorage ? .new : .active
        spools[index].updatedAt = .now
        changed()
    }

    /// Assigns a spool to a slot, moving it there and freeing whatever slot it (or the target) held.
    /// A spool can only be in one place at a time (spec §6); the target slot can hold only one spool.
    func assign(spoolID: String, to location: SpoolLocation) {
        guard let index = spools.firstIndex(where: { $0.id == spoolID }) else { return }
        // Any *other* spool currently in the target slot is bumped back to storage.
        if !location.isStorage {
            for i in spools.indices where i != index && spools[i].location.sameSlot(as: location) {
                spools[i].location = .storage
                spools[i].status = spools[i].status == .empty ? .empty : .stored
                spools[i].updatedAt = .now
            }
        }
        spools[index].location = location
        if !location.isStorage {
            spools[index].status = spools[index].remainingWeightGrams <= 0 ? .empty : .active
            if spools[index].openedAt == nil { spools[index].openedAt = .now }
        } else if spools[index].status == .active {
            spools[index].status = .stored
        }
        spools[index].updatedAt = .now
        changed()
    }

    /// Removes whatever spool sits in a slot, sending it back to storage.
    func clearSlot(_ location: SpoolLocation) {
        guard let spool = spool(at: location) else { return }
        assign(spoolID: spool.id, to: .storage)
    }

    /// When a real RFID/NFC spool is newly inserted into a slot that still holds a manually-assigned
    /// Spoolbase spool, the assignment is stale (that physical roll was taken out) — send the assigned
    /// spool back to storage so the slot shows the inserted NFC roll's own data. Only fires on the
    /// insert transition (the slot gains an NFC reading), so a deliberate later assignment is left alone.
    /// Returns a short description of each detached spool (id + slot label), so the UI can tell the user.
    @discardableResult
    func detachAssignmentsReplacedByNFC(printerSerial: String, previous: [FilamentGroup], current: [FilamentGroup]) -> [(spoolID: String, slot: String)] {
        var detached: [(spoolID: String, slot: String)] = []
        for (gi, group) in current.enumerated() {
            for (si, slot) in group.slots.enumerated() where slot.remainingWeightGrams != nil {
                let hadNFC = previous.indices.contains(gi)
                    && previous[gi].slots.indices.contains(si)
                    && previous[gi].slots[si].remainingWeightGrams != nil
                guard !hadNFC else { continue }
                let location = SpoolLocation(printerSerial: printerSerial,
                                             feeder: group.isExternal ? .ext : .ams, amsIndex: gi, slot: si)
                if let assigned = spool(at: location) {
                    let slotLabel = group.isExternal ? group.displayName : "\(group.displayName) \(slot.label)"
                    clearSlot(location)
                    detached.append((assigned.id, slotLabel))
                }
            }
        }
        return detached
    }

    // MARK: Consumption (ETAP 4 — idempotent per print job)

    /// Subtracts filament for a finished job. Idempotent: a job id already recorded is ignored, so a
    /// reconnect / restart / re-delivered event never double-subtracts (spec §17). Returns false when
    /// the job was already applied (or the spool is unknown).
    @discardableResult
    func consume(spoolID: String, grams: Double, printerSerial: String, printJobID: String) -> Bool {
        guard grams > 0 else { return false }
        guard !usageEvents.contains(where: { $0.printJobID == printJobID && $0.spoolID == spoolID }) else { return false }
        guard let index = spools.firstIndex(where: { $0.id == spoolID }) else { return false }
        spools[index].remainingWeightGrams = max(0, spools[index].remainingWeightGrams - grams)
        spools[index].totalConsumedGrams += grams
        spools[index].lastUsedAt = .now
        spools[index].updatedAt = .now
        if spools[index].remainingWeightGrams <= 0 {
            spools[index].status = .empty
            spools[index].emptiedAt = .now
        }
        usageEvents.append(SpoolUsageEvent(spoolID: spoolID, printerSerial: printerSerial,
                                           printJobID: printJobID, consumedGrams: grams))
        saveUsage()
        changed()
        return true
    }

    // MARK: Persistence

    private func changed() {
        save()
        onChange?()
    }

    private func save() { Self.write(spools, to: spoolsURL) }
    private func saveUsage() { Self.write(usageEvents, to: usageURL) }

    private static func load<T: Decodable>(_ type: T.Type, from url: URL) -> T? {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        guard let data = try? Data(contentsOf: url) else { return nil }
        return try? decoder.decode(T.self, from: data)
    }

    private static func write<T: Encodable>(_ value: T, to url: URL) {
        do {
            try FileManager.default.createDirectory(at: url.deletingLastPathComponent(), withIntermediateDirectories: true)
            let encoder = JSONEncoder()
            encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
            encoder.dateEncodingStrategy = .iso8601
            try encoder.encode(value).write(to: url, options: .atomic)
        } catch {
            NSLog("Nie można zapisać rolek Spoolbase: %@", error.localizedDescription)
        }
    }

    private static var baseDirectory: URL {
        FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask)[0]
            .appendingPathComponent("Spoolbase", isDirectory: true)
    }
    static var defaultSpoolsURL: URL { baseDirectory.appendingPathComponent("spools-v1.json") }
    static var defaultUsageURL: URL { baseDirectory.appendingPathComponent("usage-v1.json") }
}

File: windows/Gantry.Windows/Services/PhysicalSpoolStore.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text.Json;
using System.Text.Json.Serialization;
using Gantry.Models;

namespace Gantry.Services;

/// <summary>Shared Spoolbase stores so the inventory window, the AMS assign popover and the consumption
/// tracker all read/write one source of truth.</summary>
public static class SpoolbaseShared
{
    public static FilamentStore Filaments { get; } = new();
    public static PhysicalSpoolStore Spools { get; } = new();
}

/// <summary>Persists physical spools + consumption history to %AppData%\Spoolbase\ (JSON, keys shared
/// with macOS/Linux). Mirrors the Swift PhysicalSpoolStore.</summary>
public sealed class PhysicalSpoolStore
{
    private readonly List<PhysicalSpool> _spools;
    private readonly List<SpoolUsageEvent> _usage;
    private readonly string _spoolsPath;
    private readonly string _usagePath;
    private readonly string _statePath;
    private string _committed = "";
    private bool _unreadable;
    private List<string> _reviewed = new();
    public Dictionary<string, string> AccountingWarnings { get; private set; } = new();
    public string? LastError { get; private set; }
    private sealed class State
    {
        [JsonPropertyName("version")] public int Version { get; set; } = 2;
        [JsonPropertyName("spools")] public List<PhysicalSpool> Spools { get; set; } = new();
        [JsonPropertyName("usage")] public List<SpoolUsageEvent> Usage { get; set; } = new();
        [JsonPropertyName("warnings")] public Dictionary<string, string> Warnings { get; set; } = new();
        [JsonPropertyName("reviewed")] public List<string> Reviewed { get; set; } = new();
    }

    private static readonly JsonSerializerOptions Options = new()
    {
        WriteIndented = true,
        DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
        Converters = { new JsonStringEnumConverter(JsonNamingPolicy.CamelCase) }
    };

    public event Action? Changed;

    public IReadOnlyList<PhysicalSpool> Spools => _spools;

    public PhysicalSpoolStore(string? directory = null)
    {
        var dir = directory ?? Path.Combine(AppDataRoot.Folder, "Spoolbase");
        _spoolsPath = Path.Combine(dir, "spools-v1.json");
        _usagePath = Path.Combine(dir, "usage-v1.json");
        _spools = Load<List<PhysicalSpool>>(_spoolsPath) ?? new();
        _usage = Load<List<SpoolUsageEvent>>(_usagePath) ?? new();
        _statePath = Path.ChangeExtension(_spoolsPath, "state-v2.json");
        var state = Load<State>(_statePath);
        if (state is not null && state.Version == 2) Restore(state);
        else if (File.Exists(_statePath) || File.Exists(_statePath + ".bak"))
        {
            _unreadable = true; _spools.Clear(); _usage.Clear(); LastError = "Spoolbase state is unreadable";
        }
        _committed = JsonSerializer.Serialize(Snapshot(), Options);
    }

    // Lookup

    public PhysicalSpool? Spool(string id) => _spools.FirstOrDefault(s => s.Id == id);

    public PhysicalSpool? SpoolAt(SpoolLocation location) =>
        location.IsStorage ? null : _spools.FirstOrDefault(s => s.Location.SameSlot(location));

    public string NextSpoolId()
    {
        var max = _spools
            .Where(s => s.Id.StartsWith("SP-", StringComparison.Ordinal))
            .Select(s => int.TryParse(s.Id.AsSpan(3), out var n) ? n : 0)
            .DefaultIfEmpty(0).Max();
        return $"SP-{max + 1:00000}";
    }

    // Mutations

    public void Add(PhysicalSpool spool) { _spools.Add(spool); ChangedInternal(SaveSpools); }

    /// <summary>Creates <paramref name="count"/> fresh rolls of one definition, each with its own id,
    /// dropped straight into storage (spec §1: adding a filament creates one physical roll per spool). A
    /// full roll has remaining == nominal == <paramref name="weight"/>; an opened roll passes a smaller
    /// <paramref name="remaining"/>.</summary>
    public void CreateRolls(Guid definitionId, int count, double weight, double? remaining = null)
    {
        if (count <= 0) return;
        for (int i = 0; i < count; i++)
        {
            double rest = remaining ?? weight;
            _spools.Add(new PhysicalSpool
            {
                Id = NextSpoolId(),
                FilamentDefinitionId = definitionId,
                NominalWeightGrams = weight,
                RemainingWeightGrams = rest,
                Status = rest < weight ? SpoolStatus.Active : SpoolStatus.New,
                Location = SpoolLocation.Storage(),
                OpenedAt = rest < weight ? DateTime.UtcNow : null
            });
        }
        ChangedInternal(SaveSpools);
    }

    public void Delete(string id) { _spools.RemoveAll(s => s.Id == id); ChangedInternal(SaveSpools); }

    public void SetRemaining(string id, double grams)
    {
        var s = Spool(id);
        if (s is null) return;
        s.RemainingWeightGrams = Math.Max(0, Math.Min(grams, s.NominalWeightGrams));
        if (s.RemainingWeightGrams <= 0) s.Status = SpoolStatus.Empty;
        s.UpdatedAt = DateTime.UtcNow;
        ChangedInternal(SaveSpools);
    }

    /// <summary>Manual weighing (spec §5 "Skoryguj wagę"): store a freshly measured net amount and stamp
    /// the weighing date. <paramref name="tare"/> (empty-spool weight) is remembered for next time; the
    /// caller passes an already-net value (gross minus tare).</summary>
    public void CorrectWeight(string id, double netGrams, double? tare = null)
    {
        var s = Spool(id);
        if (s is null) return;
        s.RemainingWeightGrams = Math.Max(0, Math.Min(netGrams, s.NominalWeightGrams));
        if (tare is not null) s.TareGrams = tare;
        s.WeighedAt = DateTime.UtcNow;
        if (s.RemainingWeightGrams <= 0) { s.Status = SpoolStatus.Empty; s.EmptiedAt = DateTime.UtcNow; }
        else if (s.Status == SpoolStatus.Empty) s.Status = s.Location.IsStorage ? SpoolStatus.Stored : SpoolStatus.Active;
        s.UpdatedAt = DateTime.UtcNow;
        ChangedInternal(SaveSpools);
    }

    /// <summary>Reset a roll back to a full nominal amount (spec §6): the same physical roll, refilled —
    /// e.g. a spent roll swapped for a fresh one of the same product without minting a new id. Clears this
    /// roll's consumption/lifecycle so history starts over.</summary>
    public void ResetToFull(string id, double? nominal = null)
    {
        var s = Spool(id);
        if (s is null) return;
        double full = Math.Max(0, nominal ?? s.NominalWeightGrams);
        s.NominalWeightGrams = full;
        s.RemainingWeightGrams = full;
        s.TotalConsumedGrams = 0;
        s.OpenedAt = null;
        s.EmptiedAt = null;
        s.WeighedAt = null;
        s.Status = s.Location.IsStorage ? SpoolStatus.New : SpoolStatus.Active;
        s.UpdatedAt = DateTime.UtcNow;
        ChangedInternal(SaveSpools);
    }

    /// <summary>Move a spool to a slot, freeing whatever slot it or the target held (single location).</summary>
    public void Assign(string spoolId, SpoolLocation location)
    {
        var spool = Spool(spoolId);
        if (spool is null) return;
        if (!location.IsStorage)
            foreach (var other in _spools.Where(o => o.Id != spoolId && o.Location.SameSlot(location)))
            {
                other.Location = SpoolLocation.Storage();
                if (other.Status != SpoolStatus.Empty) other.Status = SpoolStatus.Stored;
                other.UpdatedAt = DateTime.UtcNow;
            }
        spool.Location = location;
        if (!location.IsStorage)
        {
            spool.Status = spool.RemainingWeightGrams <= 0 ? SpoolStatus.Empty : SpoolStatus.Active;
            spool.OpenedAt ??= DateTime.UtcNow;
        }
        else if (spool.Status == SpoolStatus.Active) spool.Status = SpoolStatus.Stored;
        spool.UpdatedAt = DateTime.UtcNow;
        ChangedInternal(SaveSpools);
    }

    /// <summary>When a real RFID/NFC spool is newly inserted into a slot that still holds a manually
    /// assigned Spoolbase spool, the assignment is stale (that roll was taken out) - send it back to
    /// storage so the slot shows the inserted NFC roll's own data. Fires only on the insert transition
    /// (the slot gains an NFC reading), so a deliberate later assignment is left alone.</summary>
    public List<(string SpoolId, string Slot)> DetachAssignmentsReplacedByNfc(string printerSerial, List<FilamentGroup> previous, List<FilamentGroup> current)
    {
        var detached = new List<(string, string)>();
        for (int gi = 0; gi < current.Count; gi++)
        {
            var group = current[gi];
            for (int si = 0; si < group.Slots.Count; si++)
            {
                if (group.Slots[si].RemainingWeightGrams is null) continue;
                bool hadNfc = gi < previous.Count && si < previous[gi].Slots.Count
                    && previous[gi].Slots[si].RemainingWeightGrams is not null;
                if (hadNfc) continue;
                var location = SpoolLocation.At(printerSerial, group.IsExternal ? SpoolFeeder.Ext : SpoolFeeder.Ams, gi, si);
                if (SpoolAt(location) is { } assigned)
                {
                    string slotLabel = group.IsExternal ? group.DisplayName : $"{group.DisplayName} {group.Slots[si].Label}";
                    Assign(assigned.Id, SpoolLocation.Storage());
                    detached.Add((assigned.Id, slotLabel));
                }
            }
        }
        return detached;
    }

    /// <summary>Idempotent per print job: a job id already recorded is ignored (no double-count).</summary>
    public bool Consume(string spoolId, double grams, string printerSerial, string printJobId)
    {
        if (grams <= 0 || _reviewed.Any(job => printJobId == job || printJobId.StartsWith(job + "#", StringComparison.Ordinal))) return false;
        if (_usage.Any(u => u.PrintJobId == printJobId && u.PrinterSerial == printerSerial)) return false;
        var s = Spool(spoolId);
        if (s is null) return false;
        s.RemainingWeightGrams = Math.Max(0, s.RemainingWeightGrams - grams);
        s.TotalConsumedGrams += grams;
        s.LastUsedAt = DateTime.UtcNow;
        s.UpdatedAt = DateTime.UtcNow;
        if (s.RemainingWeightGrams <= 0) { s.Status = SpoolStatus.Empty; s.EmptiedAt = DateTime.UtcNow; }
        _usage.Add(new SpoolUsageEvent { SpoolId = spoolId, PrinterSerial = printerSerial, PrintJobId = printJobId, ConsumedGrams = grams });
        var saved = SaveState();
        Changed?.Invoke();
        return saved;
    }

    // Persistence

    public IReadOnlyList<SpoolUsageEvent> UsageEvents => _usage;

    private void ChangedInternal(Action save) { save(); Changed?.Invoke(); }
    private void SaveSpools() => SaveState();
    private State Snapshot() => new() { Spools = _spools, Usage = _usage, Warnings = AccountingWarnings, Reviewed = _reviewed };
    private void Restore(State state)
    {
        _spools.Clear(); _spools.AddRange(state.Spools);
        _usage.Clear(); _usage.AddRange(state.Usage);
        AccountingWarnings = state.Warnings; _reviewed = state.Reviewed;
    }
    public void WarnAccounting(string job, string name) { if (_reviewed.Contains(job)) return; AccountingWarnings[job] = name; SaveState(); Changed?.Invoke(); }
    public void ClearAccountingWarnings() { _reviewed.AddRange(AccountingWarnings.Keys.Where(job => !_reviewed.Contains(job))); AccountingWarnings.Clear(); SaveState(); Changed?.Invoke(); }
    private bool SaveState()
    {
        if (_unreadable) return false;
        try
        {
            var json = JsonSerializer.Serialize(Snapshot(), Options);
            AtomicFile.WriteAllText(_statePath, json, text => Deserialize<State>(text)?.Version == 2);
            _committed = json; LastError = null;
            return true;
        }
        catch (Exception e)
        {
            Restore(JsonSerializer.Deserialize<State>(_committed, Options)!);
            LastError = e.Message;
            App.LogError("Saving Spoolbase", e);
            return false;
        }
    }

    private static T? Load<T>(string path)
    {
        T? value = default;
        // Falls back to the last good copy when a crash mid-write left the file empty or cut short.
        AtomicFile.ReadAllText(path, text => (value = Deserialize<T>(text)) is not null);
        return value;
    }

    private static T? Deserialize<T>(string text)
    {
        try { return JsonSerializer.Deserialize<T>(text, Options); }
        catch (JsonException) { return default; }
    }

}
